### ISSUE TITLE
Threadly 5.32: PriorityScheduler.workerIdle spin locking with high thread limit

### ISSUE THREAD

**Issue by veloxity-bmayer** [threadly/threadly#247]
Hi,

I'm working on performing a large test of a piece of our SaaS infrastructure using threadly as the library to handle queuing on threads. However I'm running into an issue that appears randomly where threadly will create all the threads up to the limit. There bug might be in my code, but it happens less when the threadpool limit is smaller and the stack always seems to be in PriorityScheduler$WorkerPool.workerIdle when it happens and not a task I scheduled. On some occasions when this happens threadly will correct itself, but most of the time it gets stuck at near 100% cpu usage.

![threadly_spinlock](https://user-images.githubusercontent.com/46939815/51568184-243db780-1e67-11e9-90af-518c83ef7e93.png)
![visualvm overview](https://user-images.githubusercontent.com/46939815/51568185-243db780-1e67-11e9-8207-42dfa02a3dae.png)

I have threadly configured to use 768 threads, as a lot of the work we execute is IO bound either on an external service like MySQL, WebService, Disk, Network, etc... and this is designed to host around 5,000 active sessions with a lot of background processing. The gradual thread increase is fine as that's the process starting up normally

To start diagnosing the issue I gathered a "non-standard" thread dump via an internal tool as the jvm becomes unresponsive as **kill -3** gets sent to the syslog with a lot of other information making it hard to filter out.
[threadly_non_standard_dump.txt](https://github.com/threadly/threadly/files/2784757/threadly_non_standard_dump.txt)

Doing some light processing of that shows
cat ~/out.txt | grep workerIdle\( | sort | uniq -c
- 122 threads with
   - `at - app//org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:751)`
- 382 threads with
   - `at app//org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:796)`
- 32 threads with
   - `at app//org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:804)`

These line numbers make no sense in the source code so I guessed it was being inlined by the JVM and just the last "safe point" was being reported by the thread dump tools. I did some more research about "safe points" and found the following article http://psy-lob-saw.blogspot.com/2015/12/safepoints.html. This lead me to the "-XX:CompileCommandFile" jvm argument to turn off inlining of certain methods. I used this to exclude any workerIdle methods with the following file:
`exclude * workerIdle`

That gave this thread dump
[threadly_thread_dump.txt](https://github.com/threadly/threadly/files/2784758/threadly_thread_dump.txt)
I analyzed that the same way and gave more detailed information like this:
- 318 threads with
   - `at org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:751)`
- 151 threads with
   - `at org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:777)`
- 59 threads with
   - `at org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:778)`
- 116 threads with
   - `at org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:804)`
- 91 threads with
   - `at org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:809)`
- 32 threads with
   - `at org.threadly.concurrent.PriorityScheduler$WorkerPool.workerIdle(PriorityScheduler.java:813)`

I can gather more information if needed. I've been looking at this for a while and I don't understand the state of the threads and how this happens.

Thanks.

**Comment by jentfoo** [comment]
Thank you for filing the issue, we are happy to help here.

Based off what I am seeing initially I don't believe the problem is inside threadly.  I believe what you are seeing is a safe point bias in the profiling (which can be a very challenging problem to get around in the JVM).  This is why the line numbers seem like low effort operations but which you see a lot of threads (theoretically) blocked on.

I can say that I have used this pool in a wide range of use cases, including similar ones to what you describe.  For executed tasks we have benchmarks which tend to show the contention / overhead is relatively reasonable up to 1k threads (and as work is added for the tasks, this number goes up quickly, for example 2ms of processing should allow overheads to remain low till well over 2.5k threads).  For scheduled tasks this overhead happens much quicker, but I am going to assume due to throughput that is not what you are doing here?

What I think is happening here is likely somewhere else in your application the JIT is removing safe point checks (likely a tight loop).  You can see threads which are trying to safe point to then spin lock at the native call `SafepointSynchronize::begin()`.  I believe this is likely where the threads are actually stuck inside threadly code (where the safe point checks remain).

Can you try adding the JVM flag `-XX:+UseCountedLoopSafepoints` to see if it makes any difference in the application?  This seems like a good first step, but of course will come at a little bit of a performance penalty.  If it does seem to resolve the issue, finding the tight loop where the JIT may be removing the safe point check will be the next step.  If it does not resolve the issue let me know and I will try to help more (if we are lucky it might lead to an improvement in the library).

**Comment by jentfoo** [comment]
Oh, in case it was not clear, the reason the jit removing the safe point checks in other parts of the code could cause this is because all threads must safe point before a safe point is reached.  So the threads where the check remains will just spin lock as they wait for the other threads to safe point.

**Comment by veloxity-bmayer** [comment]
There is a mix of scheduled tasks and submitted tasks. In this scenario there is around 2000-2400 scheduled tasks if I had to guess. With 1,800 of them running around midnight as low priority tasks. If that's a performance issue/concern or could result in this I can restructure those to drastically lower that number, just have put that off for now. I'm not sure about the number of submitted tasks, but if I had to guess it would be around 5,000 to 10,000 tasks submitted all together when this happens.

I've been experimenting with child pools and haven't modified my codebase to be used everywhere yet.


I added the  -XX:+UseCountedLoopSafepoints argument both with and without `exclude * workerIdle` from the compilation options and I'm encountering the same issue. I have thread dump being created right now. VisualVM takes a while to complete when this happens

**Comment by jentfoo** [comment]
In that case, I need to understand the application more.  Would you want to email me directly (jent[at]threadly.org)?  If you would prefer to keep it on github that's fine, I was just trying to help with a faster communication cycle.

**Comment by jentfoo** [comment]
@veloxity-bmayer I went ahead and merged the change we discussed over email: https://github.com/threadly/threadly/commit/11a5faf74f0bf61e1749e13d256970d6e22ecb9f

I did not PR since it was pretty simple and you already knew what it was.

The benchmarks did not show much improvement (some for scheduled tasks at high thread counts like you are using), but they also did not show any regressions in performance.  To give a bit more back story here, the `canExecute` is something that is far from ideal.  This check does cost wasted cycles, but we have not found other solutions that don't either bottleneck (ie we used to have one or sometimes two threads doing this management to hand out the tasks, so there was no thread contention here, but eventually that thread could bottleneck with tasks), or don't remove other guarantees the library provides (specifically reliable `.remove` of tasks is a large feature of threadly).

This is what spawned the `UnfairExecutor`, to get rid of some of those requirements and instead focus on just pure throughput.  Of course that is no help here because you need scheduled tasks, and your tasks are not uniform enough for that implementation.

All in all, I think this was a great change, I appreciate you bringing it to our attention.  I know from discussions in email that you have found other threadly tooling to mitigate this problem some, but let me know if you want me to go ahead and release this as 5.33.

**Comment by jentfoo** [comment]
5.33 was just released to maven central (and thus should be available soon).  Let me know if any further issues come up.
